File: Second task/secondTask.cpp
#include <iostream>
#include <map>

using namespace std;


int calculate(string line){
    map<char, int> dict = {
        {'I', 1}, {'V', 5}, {'X', 10}, 
        {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000}
    };

    // Проверяем, все ли символы допустимы
    for (char item : line){
        if (dict.find(toupper(item)) == dict.end()){
            cout << "Такого числа не существует, попробуйте еще раз:" << endl;
            return -1; 
        }
    }

    // Начинаем с последней цифры
    int result = dict.at(toupper(line[line.size() - 1]));

    // Идём справа налево 
    for (size_t i = line.size() - 1; i > 0; --i){
        if (dict.at(toupper(line[i])) <= dict.at(toupper(line[i-1]))){
            result += dict.at(toupper(line[i-1]));  // Прибавляем
        } else {
            result -= dict.at(toupper(line[i-1]));  // Вычитаем
        }
    }
    return result; 
}

int main(){
    string line;
    cout << "Введите строку: " << endl;
    cin >> line;
    int res = calculate(line);
    // Если ввели некорректное число, просим ввести заново
    while (res == -1){
        cin >> line;
        res = calculate(line);
    }
    cout << res << endl;
}

File: Second task/secondTask.java
import java.util.HashMap;  // Импорт класса HashMap
import java.util.Map;      // Импорт интерфейса Map
import java.util.Scanner;  // Импорт класса Scanner для ввода

public class Main {
    public static int calculate(String line) {
        // Создаём словарь через HashMap
        Map<Character, Integer> dict = new HashMap<>();
        dict.put('I', 1);   
        dict.put('V', 5);
        dict.put('X', 10);
        dict.put('L', 50);
        dict.put('C', 100);
        dict.put('D', 500);
        dict.put('M', 1000);
        // toCharArray() преобразует строку в массив символов
        for (char item : line.toCharArray()) {
            // Character.toUpperCase() в верхний регистр
            // containsKey() - проверяет наличие ключа
            if (!dict.containsKey(Character.toUpperCase(item))) {
                System.out.println("Такого числа не существует, попробуйте еще раз:");
                return -1;  // Код ошибки
            }
        }
        
        int result = dict.get(Character.toUpperCase(line.charAt(line.length() - 1)));
        
        for (int i = line.length() - 1; i > 0; i--) {
            if (dict.get(Character.toUpperCase(line.charAt(i))) <= 
                dict.get(Character.toUpperCase(line.charAt(i - 1)))) {
                result += dict.get(Character.toUpperCase(line.charAt(i - 1)));
            } else {
                result -= dict.get(Character.toUpperCase(line.charAt(i - 1)));
            }
        }
        return result;
    }
    

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);  // Создаём сканер для ввода
        System.out.println("Введите строку: ");
        String line = scanner.next();  // Читаем слово
        int res = calculate(line);
        
        while (res == -1) {
            line = scanner.next();
            res = calculate(line);
        }
        System.out.println(res);
        scanner.close();  // Закрываем сканер
    }
}